File: scripts/load-dataset.js
// Loads data/queries.csv into the SQLite primary store.
//
// Usage:  npm run load          (after `npm run generate`)
//         npm run setup         (generate + load in one step)
//
// To load YOUR OWN dataset instead, just place a CSV with a `query,count`
// header at data/queries.csv (or set DATASET path) and run this script.

import fs from 'node:fs';
import readline from 'node:readline';
import { config } from '../src/config.js';
import { initDb, bulkInsert, totalRows, closeDb } from '../src/db.js';

const CHUNK = 5000; // rows per transaction

async function main() {
  const file = process.env.DATASET || config.paths.dataset;
  if (!fs.existsSync(file)) {
    console.error(`Dataset not found: ${file}`);
    console.error('Run `npm run generate` first, or place a CSV there.');
    process.exit(1);
  }

  initDb();
  const rl = readline.createInterface({
    input: fs.createReadStream(file),
    crlfDelay: Infinity,
  });

  let isHeader = true;
  let buffer = [];
  let loaded = 0;
  let skipped = 0;
  const t0 = Date.now();

  const flush = () => {
    if (buffer.length) {
      bulkInsert(buffer);
      loaded += buffer.length;
      buffer = [];
      process.stdout.write(`\r  loaded ${loaded.toLocaleString()} rows...`);
    }
  };

  for await (const line of rl) {
    if (isHeader) {
      isHeader = false;
      // tolerate a header row; if it doesn't look like one, treat as data
      if (/query\s*,\s*count/i.test(line)) continue;
    }
    const idx = line.lastIndexOf(',');
    if (idx === -1) {
      skipped++;
      continue;
    }
    const query = line.slice(0, idx).trim().toLowerCase();
    const count = parseInt(line.slice(idx + 1).trim(), 10);
    if (!query || !Number.isFinite(count)) {
      skipped++;
      continue;
    }
    buffer.push({ query, count });
    if (buffer.length >= CHUNK) flush();
  }
  flush();

  const total = totalRows();
  closeDb();

  const secs = ((Date.now() - t0) / 1000).toFixed(1);
  process.stdout.write('\n');
  console.log(`Done in ${secs}s`);
  console.log(`  rows loaded:   ${loaded.toLocaleString()}`);
  console.log(`  rows skipped:  ${skipped.toLocaleString()}`);
  console.log(`  rows in table: ${total.toLocaleString()}`);
  if (total < 100_000) {
    console.warn(
      `  NOTE: table has < 100k rows. Increase DATASET_SIZE and re-run setup.`
    );
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scripts/load_dataset.py
"""Load data/queries.csv into the SQLite store (the primary store).

Idempotent: re-running replaces counts for existing queries. Run after
scripts/generate_dataset.py. The app builds its in-memory trie from this table
at startup.
"""
from __future__ import annotations

import csv
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from app.config import config  # noqa: E402
from app.db import Database  # noqa: E402


def _rows(path: Path):
    with path.open(newline="") as f:
        reader = csv.reader(f)
        header = next(reader, None)
        if header != ["query", "count"]:
            raise SystemExit(f"unexpected CSV header: {header!r} (expected query,count)")
        for row in reader:
            if len(row) != 2:
                continue
            q, c = row[0].strip(), row[1].strip()
            if not q:
                continue
            try:
                yield q, int(c)
            except ValueError:
                continue


def main() -> None:
    dataset = config.paths.dataset
    if not dataset.exists():
        raise SystemExit(
            f"dataset not found at {dataset}. Run scripts/generate_dataset.py first."
        )
    db = Database()
    t0 = time.time()
    n = db.bulk_upsert(_rows(dataset))
    dt = time.time() - t0
    print(f"Loaded {n:,} rows into {config.paths.db} in {dt:.1f}s")
    print(f"  total distinct queries in store: {db.row_count():,}")
    db.close()


if __name__ == "__main__":
    main()
